go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,367 @@
+// Copyright 2016, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package planbuilder
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/youtube/vitess/go/vt/sqlparser"
+)
+
+// PlanBuilder represents any object that's used to
+// build a plan. The top-level PlanBuilder will be a
+// tree that points to other PlanBuilder objects.
+// Currently, JoinBuilder and RouteBuilder are the
+// only two supported PlanBuilder objects. More will be
+// added as we extend the functionality.
+// Each Builder object builds a Plan object, and they
+// will mirror the same tree. Once all the plans are built,
+// the builder objects will be discarded, and only
+// the Plan objects will remain.
+type PlanBuilder interface {
+	// Order is a number that signifies execution order.
+	// A lower Order number Route is executed before a
+	// higher one. For a node that contains other nodes,
+	// the Order represents the highest order of the leaf
+	// nodes.
+	Order() int
+}
+
+// JoinBuilder is used to build a Join primitive.
+// It's used to buid a normal join or a left join
+// operation.
+// TODO(sougou): struct is incomplete.
+type JoinBuilder struct {
+	// IsLeft is true if the operation is a left join.
+	IsLeft bool
+	order  int
+	// Left and Right are the nodes for the join.
+	Left, Right PlanBuilder
+}
+
+// Order returns the order of the node.
+func (jb *JoinBuilder) Order() int {
+	return jb.order
+}
+
+// MarshalJSON marshals JoinBuilder into a readable form.
+// It's used for testing and diagnostics. The representation
```

Comment:
Thanks for adding. It's very helpful to know it's not for production code but just diagnostics.